compiler/optimizer/VPHandlers.cpp

```diff
@@ -8321,67 +8321,195 @@ void replaceWithSmallerType(OMR::ValuePropagation *vp, TR::Node *node)
+   // and so it preserves signed ordering, i.e. if x, y are possible values of
+   // the child with x <= y, then truncate(x) <= truncate(y).
+   //
+   // To see why this must be the case, let b=resultSizeBits, let Q=lowTopBits,
+   // and let x be the signed value of the child. Note that signBit = 2**(b-1).
+   // Then
+   //
+   //    Q * 2**b <= low + 2**(b-1)
+   //    Q * 2**b - 2**(b-1) <= low
+   //
+   // and
+   //
+   //    high + 2**(b-1) < (Q+1) * 2**b
+   //    high < (Q+1) * 2**b - 2**(b-1) = Q * 2**b + 2**(b-1)
+   //
+   // Since low <= x <= high, we know that
+   //
+   //    Q * 2**b - 2**(b-1) <= low <= x <= high < Q * 2**b + 2**(b-1)
+   //    -2**(b-1) <= x - Q * 2**b < 2**(b-1)
+   //
+   // Since [-2**(b-1), 2**(b-1)) is the signed range of the result type, the
+   // signed value of the result is x - Q * 2**b for every x.
+   //
+   // So if x, y are possible signed child values with x <= y, then
+   //
+   //    truncate(x) = x - Q * 2**b <= y - Q * 2**b = truncate(y),
+   //
+   // and in particular since low, high are also possible signed child values
+   // with low <= x <= high, we also know that
+   //
+   //    truncate(low) <= truncate(x) <= truncate(high).
+   //
+   // Therefore, truncating low and high will produce a signed range containing
+   // all possible truncated results.
+   //
+   // NOTE: The above proof assumes that the addition of signBit in the biased
+   // endpoints calculation does not signed-overflow, i.e. that the signed
+   // value of the 64-bit sum is ({low,high} + 2**(b-1)). But it is possible
+   // for such overflow to have occurred in both calculations. If it did, then
+   // the signed values of the sums were really
+   //
+   //    {low,high} + 2**(b-1) - 2**64.
+   //
+   // In this case, we can simply take Q = topLowBits + 2**(64-b) and
+   // everything else will still follow.
+   //
+   // With this justification, get the truncated (and re-extended) bounds. Cast
+   // to unsigned before shifting left to avoid undefined behaviour.
+   //
+   const uint64_t origDiff = (uint64_t)high - (uint64_t)low;
+   low = (int64_t)((uint64_t)low << padding64) >> padding64;
+   high = (int64_t)((uint64_t)high << padding64) >> padding64;
+
+   TR_ASSERT_FATAL_WITH_NODE(node, min <= low, "truncated lower bound is too low");
+   TR_ASSERT_FATAL_WITH_NODE(node, low <= high, "truncated bounds are out of order");
+   TR_ASSERT_FATAL_WITH_NODE(node, high <= max, "truncated upper bound is too high");
+   TR_ASSERT_FATAL_WITH_NODE(
+      node,
+      (uint64_t)high - (uint64_t)low == origDiff,
+      "truncated range is not the same size as the original range");
+
+   if (low >= 0)
+      node->setIsNonNegative(true);
+
+   // Constrain the result.
+   switch (node->getDataType())
+      {
+      case TR::Int16:
+         constraint = TR::VPShortRange::create(vp, (int16_t)low, (int16_t)high);
+         break;
+
+      case TR::Int8: // byte nodes are still using Int constraint
```

Comment:
This is an odd state of affairs that I wasn't aware of, i.e. byte nodes using int constraints. But I realize this was true before this change from you anyway. I did double check that `replaceByConstant` obtained the constraint for byte node/case as an integer one.